pints/tests/test_slice_stepout.py

```diff
@@ -0,0 +1,397 @@
+        with self.assertRaises(ValueError):
+            mcmc.set_prob_overrelaxed(-1)
+        with self.assertRaises(ValueError):
+            mcmc.set_prob_overrelaxed(4)
+
+        # Test set_a
+        mcmc.set_a(40)
+        self.assertEqual(mcmc._a, 40)
+        with self.assertRaises(ValueError):
+            mcmc.set_prob_overrelaxed(-1)
+
+        # Test get_w
+        self.assertTrue(np.all(mcmc.get_w() == np.array([2, 2])))
+
+        # Test get_m
+        self.assertEqual(mcmc.get_m(), 3.)
+
+        # Test current_log_pdf
+        self.assertEqual(mcmc.get_current_log_pdf(), mcmc._current_log_pdf)
+
+        # Test current_slice height
+        self.assertEqual(mcmc.get_current_slice_height(), mcmc._current_log_y)
+
+        # Test number of hyperparameters
+        self.assertEqual(mcmc.n_hyper_parameters(), 4)
+
+        # Test setting hyperparameters
+        mcmc.set_hyper_parameters([3, 100, .7, 50])
+        self.assertTrue((np.all(mcmc._w == np.array([3, 3]))))
+        self.assertEqual(mcmc._m, 100)
+        self.assertEqual(mcmc._prob_overrelaxed, 0.7)
+        self.assertEqual(mcmc._a, 50)
+
+    def test_logistic(self):
+        """
+        Test sampler on a logistic task.
+        """
+        # Load a forward model
+        model = toy.LogisticModel()
+
+        # Create some toy data
+        real_parameters = [0.015, 500]
+        times = np.linspace(0, 1000, 1000)
+        org_values = model.simulate(real_parameters, times)
+
+        # Add noise
+        noise = 10
+        values = org_values + np.random.normal(0, noise, org_values.shape)
+        real_parameters = np.array(real_parameters + [noise])
+
+        # Create an object with links to the model and time series
+        problem = pints.SingleOutputProblem(model, times, values)
+
+        # Create a log-likelihood function (adds an extra parameter!)
+        log_likelihood = pints.GaussianLogLikelihood(problem)
+
+        # Create a uniform prior over both the parameters and the new
+        # noise variable
+        log_prior = pints.UniformLogPrior(
+            [0.01, 400, noise * 0.1],
+            [0.02, 600, noise * 100],
+        )
+
+        # Create a posterior log-likelihood (log(likelihood * prior))
+        log_posterior = pints.LogPosterior(log_likelihood, log_prior)
+
+        # Choose starting points for 3 mcmc chains
+        num_chains = 1
+        xs = [real_parameters * (1 + 0.1 * np.random.rand())]
+
+        # Create mcmc routine
+        mcmc = pints.MCMCController(
+            log_posterior, num_chains, xs, method=pints.SliceStepoutMCMC)
+
+        for sampler in mcmc.samplers():
+            sampler.set_w(0.1)
+
+        # Add stopping criterion
+        mcmc.set_max_iterations(100)
+
+        # Set up modest logging
+        mcmc.set_log_to_screen(True)
+        mcmc.set_log_interval(500)
+
+        # Run!
+        print('Running...')
+        mcmc.run()
+        print('Done!')
+
```

Comment:
I'm not sure I follow what these last lines of the testing procedures are doing? They are checking that the samplers run but not sure what else they do? Do we want to check the length of the samples that are outputted for example? And the means of the outputted samples?

Reply:
Hi @ben18785 - As initially I was having some issues running the samplers on the logistic model (I don't remember what the issue was), I implemented the logistic problem as a test to check that everything was running properly. This test doesn't check anything else.

Reply:
The ideal way to write unit tests is to have them test very small bits of code (ideally in different situations), and then check that the output is what you expect. The idea being that if all the little bits function well individually, the whole should work well too. So if you encounter some issues in a particular problem, the trick is to recreate that issue in a very localised way: set up the exact situation that made it break (with very little overhead), and then re-run only the bit where the error occurred.

So instead of re-running the whole sampling routine as a test you would
1. Isolate what the problem is
2. Find some way to recreate that problem that has waaaay less overhead
3. Write a test (that fails)
4. Fix the bug

Reply:
So in this case you'd ideally have a test where you brought the sampler into some state that you know used to break it, then do a few steps via ask-and-tell, check the output is correct. No controller involved!